cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,880 @@
+        ...     (Z-1, Z),
+        ...     (X, Y),
+        ...     (X, Z),
+        ...     (Y, Z),
+        ... ]
+
+        :param adjacency_matrices: A dictionary of adjacency matrices. Keys are the time delta.
+        :param variable_names: A list of variable names. If not provided, the variable names are integers starting from 0.
+            Node names must correspond to the variable names and must not contain the lag.
+        :return: A time series causal graph.
+        """
+        assert isinstance(adjacency_matrices, dict)
+        # keys must be integers or str that can be converted to integers
+        assert all(isinstance(key, (int, str)) and int(key) == key for key in adjacency_matrices)
+
+        # get the shape of the adjacency matrices (get the first key as 0 may not be present)
+        shape = adjacency_matrices[next(iter(adjacency_matrices))].shape
+
+        if variable_names is not None:
+            variable_names_str: List[Union[str, int]] = []
+            assert len(variable_names) == shape[0], (
+                'The number of variable names must be equal to the number of nodes in the adjacency matrix.'
+                f'Got {len(variable_names)} variable names and {shape[0]} nodes.'
+            )
+            # convert the variable names to strings if they are not strings
+            for variable_name in variable_names:
+                if isinstance(variable_name, HasIdentifier):
+                    variable_names_str.append(variable_name.identifier)
+                else:
+                    variable_names_str.append(variable_name)
+
+        else:
+            variable_names_str = [f'node_{i}' for i in range(shape[0])]
+
+        # we could create the full adjacency matrix from the adjacency matrices by stacking them according to the time delta
+        # but if we have many time deltas, this could be very memory intensive
+        # so we create the graph by adding the edges one by one for each time delta
+
+        # create the empty graph
+        tsgraph = TimeSeriesCausalGraph()
+
+        for time_delta, adjacency_matrix in adjacency_matrices.items():
+            # create the edges
+            edges: List[Tuple[str, str]] = []
+            # get the edges from the adjacency matrix by getting the indices of the non-zero elements
+            for row, column in zip(*numpy.where(adjacency_matrix)):
+                edges.append(
+                    (
+                        get_name_with_lag(variable_names_str[row], time_delta),
+                        variable_names_str[column],
+                    )
+                )
+            # add the edges to the graph
+            tsgraph.add_edges_from(edges)   # type: ignore
+
+        return tsgraph
+
+    @property
+    def adjacency_matrices(self) -> Dict[int, numpy.ndarray]:
+        """
+        Return the adjacency matrix dictionary of the minimal causal graph.
+
+        The keys are the time deltas and the values are the adjacency matrices.
```

Comment:
Format as parameter docstrings for clarity

Reply:
There are no parameters. Don't think this needs any formatting.